Start EB

WeiqunZhang opened the pull request:
This is still work in progress. You can build EB and create EBFArrayBoxFactory objects from python.

WeiqunZhang committed:
Start EB

This is still work in progress. You can build EB and create
EBFArrayBoxFactory objects from python.

ax3l commented on src/EB/EBFabFactory.cpp:
```diff
@@ -0,0 +1,32 @@
+            return makeEBFabFactory(geom, ba, dm, ngrow, support);
+        },
+        py::arg("geom"), py::arg("ba"), py::arg("dm"), py::arg("ngrow"),
+        py::arg("support").
```
```suggestion
        py::arg("support"),
```

ax3l replied:
fixed.

ax3l commented on docs/source/usage/api.rst:
```diff
@@ -280,3 +280,19 @@ This is for the legacy, AoS + SoA particle containers only:
+Embedded Boundaries
+-------------------
+
+Embedded boundary (EB) support is still experimental. To build pyAMReX with
```
Although we use EBs in WarpX very well with Python... Maybe one can skip the first sentence completely or rephrase to:
```suggestion
Embedded boundary (EB) support in pyAMReX is still minimal. To build pyAMReX with
```

We only need more EB bindings if we build and allocate EBs from pyAMReX. If we just interact with EB-enabled simulations, we are already quite good (e.g., `iMultiFab` et al. is what we need for WarpX with EB).

ax3l replied:
Updated

ax3l committed:
Fix

ax3l committed:
CMake: Move EB Control in

Just simplifies the call site.

ax3l committed:
Docs Update

Signed-off-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l reviewed (approved):
Thank you for starting this! :)

ax3l committed:
Copyright File Headers

ax3l committed:
Modern Enum

ax3l committed:
Start EB Test Template

ax3l commented on tests/test_eb.py:
```diff
@@ -0,0 +1,14 @@
+
+    # TODO:
+    # EB2_Build(...)
+    # makeEBFabFactory(...)
```
@WeiqunZhang do you like to finalize your PR by adding a test that uses your new functionality, so we have test coverage?
```suggestion
    ... = EB2_Build(...)
    ... = makeEBFabFactory(...)
```

WeiqunZhang replied:
Yes, I will do it.

WeiqunZhang committed:
add ifdef AMREX_UEE_EB

WeiqunZhang committed:
Add test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

WeiqunZhang committed:
add cmake build option to setup.py

ax3l commented on src/EB/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+    if (D EQUAL 1)
+        return()
+    endif()
+
```
Ah I see it. `foreach` is not a function :)

ax3l committed:
Fix: CMake Logic

`foreach` is not a function, so once 1D was built
all EB `.cpp` files were skipped.

ax3l committed:
Simplify

Guards not needed.

ax3l reviewed (approved):
Good start, thank you! :sparkles:

ax3l commented on docs/source/usage/api.rst:
```diff
@@ -280,3 +280,19 @@ This is for the legacy, AoS + SoA particle containers only:
+Embedded Boundaries
+-------------------
+
+Embedded boundary (EB) support in pyAMReX is still minimal. To build pyAMReX with
+EB support, you need to add ``-DAMReX_EB=ON`` to CMake build options.
+
+.. autofunction:: amrex.space3d.EB2_Build
+
+.. autoclass:: amrex.space3d.EBFArrayBoxFactory
+   :members:
```
Hm, gotta build locally but something broke the autodocs for all types
https://pyamrex.readthedocs.io/en/latest/usage/api.html

ax3l replied:
I built locally and the issue is https://github.com/sizmailov/pybind11-stubgen/pull/275

`EBFArrayBoxFactory` depends on `FabFactory_FArrayBox` but pybind11-stubgen does falsly an alphabetical sorting. Hoping this PR gets accepted and unbreaks it, otherwise we need to fork pybind11-stubgen.

ax3l replied:
#509
